cloud/scope/powervs_cluster.go

```diff
@@ -1627,71 +1648,73 @@ func (s *PowerVSClusterScope) ReconcileTransitGateway() (bool, error) {
 		if err != nil {
 			return false, err
 		}
-		requeue, err := s.checkTransitGateway(tg.ID)
+		requeue, _, _, err := s.checkAndUpdateTransitGateway(tg, false)
 		if err != nil {
 			return false, err
 		}
 		return requeue, nil
 	}
 
 	// check transit gateway exist in cloud
-	tgID, requeue, err := s.isTransitGatewayExists()
+	tgID, requeue, powerVSConn, vpcConn, err := s.isTransitGatewayExists()
 	if err != nil {
 		return false, err
 	}
 	if tgID != "" {
-		s.V(3).Info("Transit gateway found in IBM Cloud")
-		s.SetStatus(infrav1beta2.ResourceTypeTransitGateway, infrav1beta2.ResourceReference{ID: &tgID, ControllerCreated: ptr.To(false)})
+		s.SetTransitGatewayStatus(&tgID, ptr.To(false), powerVSConn, vpcConn)
 		return requeue, nil
 	}
 	// create transit gateway
 	s.V(3).Info("Creating transit gateway")
-	transitGatewayID, err := s.createTransitGateway()
+	transitGatewayID, powerVSConnID, vpcConnID, err := s.createTransitGateway()
 	if err != nil {
 		return false, fmt.Errorf("failed to create transit gateway: %v", err)
 	}
 	if transitGatewayID != nil {
 		s.Info("Created transit gateway", "id", transitGatewayID)
-		s.SetStatus(infrav1beta2.ResourceTypeTransitGateway, infrav1beta2.ResourceReference{ID: transitGatewayID, ControllerCreated: ptr.To(true)})
+		s.SetTransitGatewayStatus(transitGatewayID, ptr.To(true), &infrav1beta2.ResourceReference{ID: powerVSConnID, ControllerCreated: ptr.To(true)}, &infrav1beta2.ResourceReference{ID: vpcConnID, ControllerCreated: ptr.To(true)})
 	}
 	return true, nil
 }
 
-// checkTransitGateway checks transit gateway exist in cloud.
-func (s *PowerVSClusterScope) isTransitGatewayExists() (string, bool, error) {
+// isTransitGatewayExists checks transit gateway exist in cloud and if exist with unattached connections, it attaches the connection too.
+func (s *PowerVSClusterScope) isTransitGatewayExists() (string, bool, *infrav1beta2.ResourceReference, *infrav1beta2.ResourceReference, error) {
 	// TODO(karthik-k-n): Support regex
-	transitGateway, err := s.TransitGatewayClient.GetTransitGatewayByName(*s.GetServiceName(infrav1beta2.ResourceTypeTransitGateway))
-	if err != nil {
-		return "", false, err
-	}
-	if transitGateway == nil || transitGateway.ID == nil {
-		s.V(3).Info("Transit gateway not found in IBM Cloud")
-		return "", false, nil
+	var transitGateway *tgapiv1.TransitGateway
+	var err error
+
+	if s.IBMPowerVSCluster.Spec.TransitGateway != nil && s.IBMPowerVSCluster.Spec.TransitGateway.ID != nil {
+		transitGateway, _, err = s.TransitGatewayClient.GetTransitGateway(&tgapiv1.GetTransitGatewayOptions{
+			ID: s.IBMPowerVSCluster.Spec.TransitGateway.ID,
+		})
+	} else {
+		transitGateway, err = s.TransitGatewayClient.GetTransitGatewayByName(*s.GetServiceName(infrav1beta2.ResourceTypeTransitGateway))
 	}
-	requeue, err := s.checkTransitGateway(transitGateway.ID)
+
 	if err != nil {
-		return "", false, err
+		return "", false, nil, nil, err
 	}
-	return *transitGateway.ID, requeue, nil
-}
 
```

Comment:
In a case where `spec.TG.ID` is `nil` and given TG name doesn't exist in cloud, `GetTransitGatewayByName`, returns `nil`. Without a nil check here, it would fail going further at `checkTransitGatewayStatus` where `*tg.Status` would be nil right?  @dharaneeshvrd, Please correct me if my understanding is not right..